Add rule to migrate away from ktlint-disable directives

paul-dingemans opened the pull request:
## Description

Add rule to migrate ktlint-disable directives from comments to @Suppress or @SuppressWarnings annotations

Note that the SuppressionLocatorBuilder never ignore this rule as otherwise the suppression hint to ignore *all* rules would also ignore the rule which has to migrate this suppression hint.

REVIEW suggestion:
* The functional change for #1947 is contained in [this commit](https://github.com/pinterest/ktlint/pull/2068/commits/428d52e637d931754b3756f8912f64cea884f8d5)
* Other commits are supportive to the change

Closes #1947

## Checklist

<!-- Following checklist may be skipped in some cases -->
- [X] PR description added
- [X] tests are added
- [X] KtLint has been applied on source code itself and violations are fixed
- [X] [documentation](https://pinterest.github.io/ktlint/) is updated
- [X] `CHANGELOG.md` is updated

In case of adding a new rule:
- [ ] Rule is added to [rules documentation](https://pinterest.github.io/ktlint/rules/standard/)

paul-dingemans commented:
@atulgpt I you also want to contribute by reviewing changes, you're welcome to review this PR. You can review it in detail or just briefly. Normally, I would just merge the change as there are almost no regular contributors on the project that would like to do this. I you don't want to review, just let me know.

REVIEW suggestion:
* The functional change for is contained in [this commit](https://github.com/pinterest/ktlint/pull/2068/commits/428d52e637d931754b3756f8912f64cea884f8d5)
* Other commits are supportive to the change

atulgpt commented on ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRule.kt:
```diff
@@ -0,0 +1,466 @@
+ * A ktlint-disable directive is replaced with an annotation on the closest parent declaration or expression, or as annotation on the file
+ * level in case the directive is associated with a top level element.
+ *
+ * If the target element is annotated with a [Suppress], or if missing with a [SuppressWarnings] annotation then the ktlint-directive will
```
Will `if missing` be `annotated with`?

paul-dingemans replied:
I will change the sentence to `If the target element is annotated with a [Suppress] (or, if missing, is annotated with a [SuppressWarnings] annotation) then the ktlint-directive will be matched against this annotation.`. Hopefully that is more clear.

atulgpt replied:
That is clearer than above

atulgpt commented on ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRule.kt:
```diff
@@ -0,0 +1,466 @@
+ *
+ * If the target element is not annotated with [Suppress] or [SuppressWarnings] then a [Suppress] annotation is added.
+ *
+ * A ktlint-enable directive is removed as annotations have a scope in which the suppression will be active.
```
Should `directive` be `directives`?

paul-dingemans replied:
I will change the sentence to `Ktlint-enable directives are removed as annotations have a scope in which the suppression will be active.`.

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+        fun `Given a @file Suppress annotation`() {
+            val code =
+                """
+                @file:Suppress("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar")
```
We can also test `@file:[Suppress("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar") SuppressWarnings("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar")]`

atulgpt replied:
Few more variations
```
@file:[Suppress(names = arrayOf("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar"))]
@file:[Suppress(names = ["ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar"])]
@file:Suppress(names = arrayOf("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar"))
@file:Suppress(names = ["ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar"])
```

atulgpt replied:
I think you can add two more `@file:[Suppress("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar"), SomeOtherAnnotation]` or @file:[Suppress("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar"), SuppressWarnings]`

`@file:[AAA, Suppress("ktlint:bar", "ktlint:standard:foo", "ktlint:foo-bar")]` (here AAA is alphabetically sorted first annotation)

paul-dingemans replied:
Those tests won't add value once you know how the rule actually works. So from the unit test perspective those test won't increase the line coverage. The order of the annotations within the array is for example totally irrelevant for the logic. Also, it no longer makes a difference whether the annotation is included in a array or starts with the `@` when not included in an annotation array.

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+        }
+
+        @Test
+        fun `Given an EOL comment with a ktlint-disable directive on an import and an existing @file Suppress annotation`() {
```
One more TC(although very edge case)
```
val code =
    """
     @file:Suppress
     import foo.bar
     import foobar.* // ktlint-disable no-wildcard-imports
    """
```

paul-dingemans replied:
Done, with example below:
```
@file:Suppress

import foobar.* // ktlint-disable no-wildcard-imports
```

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+                """.trimIndent()
+            val formattedCode =
+                """
+                @file:Suppress("aaa", "ktlint:standard:no-wildcard-imports", "zzz")
```
do we have a wrapping rule around annotation? If yes then we should also add one TC which after autocorrect violates the wrapping rule

paul-dingemans replied:
Annotations are wrapped similar to any function call. The wrapping rule is ran later in the process. I see no need to test the combination explicitly.

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+                package foo
+
+                import foo.bar
+                import foobar.* // ktlint-disable no-wildcard-imports
```
Does it make sense to add this TC `import foobar.* // ktlint-disable no-wildcard-imports // some comment`?

paul-dingemans replied:
Lol, you're really creative. The SuppressionLocatorBuilder will print warnings but processes it anyways. i am not in favor to try to be 100% fool proof. I expect a user to inspect format fixes proposed by ktlint before committing.

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+        }
+
+        @Test
+        fun `Given a SuppressWarnings annotation on a declaration`() {
```
Should we add TC around
```
fun foo(
    a: Int /* ktlint-supression */
    b: Int // ktlint-suppression
)
```

paul-dingemans replied:
> > @atulgpt I you also want to contribute by reviewing changes, you're welcome to review this PR. You can review it in detail or just briefly. Normally, I would just merge the change as there are almost no regular contributors on the project that would like to do this. I you don't want to review, just let me know.
> > REVIEW suggestion:
> > ```
> > * The functional change for is contained in [this commit](https://github.com/pinterest/ktlint/pull/2068/commits/428d52e637d931754b3756f8912f64cea884f8d5)
> > 
> > * Other commits are supportive to the change
> > ```
> 
> Hi, @paul-dingemans I have added a few more TC suggestions. Will take a look in the logic as well

Wow, I am really amazed at the good quality of the review comments. Tnx!  I have left some review comments unanswered. I have to think about them.

paul-dingemans replied:
Good find. With latest change, the slightly rewritten example below:
```
fun foobar(
    foo: Int, /* ktlint-disable rule-foo */
    bar: Int, // ktlint-disable rule-bar
)
```
is changed to:
```
@Suppress("ktlint:standard:rule-bar", "ktlint:standard:rule-foo")
fun foobar(
    foo: Int,
    bar: Int,
)
```
which is not correct. It should have been:
```
@file:Suppress("ktlint:standard:rule-foo")

@Suppress("ktlint:standard:rule-bar")
fun foobar(
    foo: Int,
    bar: Int,
)
```

But I am doubting whether it is worth the effort to defend against this. A more realistic and troublesome example is:
```
fun foo() {
    /* ktlint-disable max-line-length */
}

fun bar() {
   val someLonnnnnnnnnnnnnnnnnnnnnnnnnnnnngIdentifierExceedingMaxLineLength = "bar"
}
```
The ktlint-disable directive does not have a logical scope (e.g. not bound to an ASTNode) but just to line number. I think that in such case, autocorrect is going to be difficult. Assuming that devs won't do stuff like this, I propose to do following:
* if a matching ktlint-enable directive in a block comment is found at a sibling node of the ktlint-disable directive (also in a block comment) then it is propagate to the first parent declaration/expression in same way as a disable directive in an EOL comment.
* Otherwise the ktlint-disable directive is just reported as an error which can not be autocorrected.

Example 1:
```
fun foobar(
     /* ktlint-disable rule-foo */
    foo: Int,
    bar: Int, // ktlint-disable rule-bar
    /* ktlint-enable rule-foo */
) {
    // do something
}
```
will be changed to:
```
@Suppress("ktlint:standard:rule-bar", "ktlint:standard:rule-foo")
fun foobar(
    foo: Int,
    bar: Int,
) {
    // do something
}
```

Example 2:
```
fun foobar(
     /* ktlint-disable rule-foo */
    foo: Int,
    bar: Int, // ktlint-disable rule-bar
) {
    /* ktlint-enable rule-foo */
    // do something
}
```
will result in a violation error which can not be autocorrected, and the code being changed to
```
@Suppress("ktlint:standard:rule-bar")
fun foobar(
     /* ktlint-disable rule-foo */
    foo: Int,
    bar: Int,
) {
    // do something
}
```

atulgpt replied:
Make sense. As no one would be using out-of-place enable and disable directives. Let me know once you update the final code. I would like to take a look and obviously will gaining lots of understanding around these

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+        fun `Given a ktlint-disable directive with rule-id prefixed with a rule set id`() {
+            val code =
+                """
+                /* ktlint-disable some-rule-set:some-rule-id */
```
One TC can be added around with multiple `ktlint-disable`
```
/* ktlint-disable some-rule-set:some-rule-id-1 */
/* ktlint-disable some-rule-set:some-rule-id-2 */

/* ktlint-disable some-rule-set:some-rule-id-3 */
```

paul-dingemans replied:
This already existed in TC `Given multiple ktlint-disable directives which have to merged into an existing @file Suppress annotation`

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+        fun `Given a ktlint-disable directive without rule-id`() {
+            val code =
+                """
+                fun foo() {
```
what will be behaviour when `/* ktlint-disable */` is added to the `init { }` block?

paul-dingemans replied:
Given code below:
```
class Foo() {
    var foo: String
    var bar: String

    init {
        /* ktlint-disable standard:foo */
        foo = "foo"
    }

    init { // ktlint-disable standard:bar
        bar = "bar"
    }
}
```
this results in adding a @Suppress to both init blocks which is not valid code. I will add the test case and fix. Example above uncovered that after an initial fix, the @Suppress annotation on classes and functions is not inserted correctly in the AST resulting in two distinct annotation on top of the class. Need to do some refactoring there (not yet pushed).

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+                """.trimIndent()
+            val formattedCode =
+                """
+                @Suppress("ktlint:standard:some-rule-id")
```
Are we also checking `experimental` rules and prefixing with `experimental` instead of `standard`?

atulgpt commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+            val code =
+                """
+                fun foo() {
+                    /* ktlint-disable some-rule-set:some-rule-id */
```
multiple enable disable directives either like below
```
/* ktlint-disable some-rule-set:some-rule-id-1 */
/* ktlint-disable some-rule-set:some-rule-id-2 */
// code
/* ktlint-enable some-rule-set:some-rule-id-1 */
/* ktlint-enable some-rule-set:some-rule-id-2 */
```
or
```
/* ktlint-disable some-rule-set:some-rule-id-1 */
// code 1
/* ktlint-enable some-rule-set:some-rule-id-1 */

/* ktlint-disable some-rule-set:some-rule-id-2 */
// code 2
/* ktlint-enable some-rule-set:some-rule-id-2 */
```

atulgpt replied:
Hi, @paul-dingemans here expected output would be `@Suppress("ktlint:some-rule-set:some-rule-id-1", "ktlint-disable some-rule-set:some-rule-id-2")` so basically I thing we have to collect all the suppression in the scope and add those in the single `@Suppress`

paul-dingemans replied:
Yes indeed they should be collected in a single @Suppress. Will have to fix this.

atulgpt commented:
> @atulgpt I you also want to contribute by reviewing changes, you're welcome to review this PR. You can review it in detail or just briefly. Normally, I would just merge the change as there are almost no regular contributors on the project that would like to do this. I you don't want to review, just let me know.
> 
> REVIEW suggestion:
> 
>     * The functional change for is contained in [this commit](https://github.com/pinterest/ktlint/pull/2068/commits/428d52e637d931754b3756f8912f64cea884f8d5)
> 
>     * Other commits are supportive to the change

Hi, @paul-dingemans I have added a few more TC suggestions. Will take a look in the logic as well

paul-dingemans commented on ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/KtlintSuppressionRuleTest.kt:
```diff
@@ -0,0 +1,798 @@
+                    LintViolation(1, 61, "Identifier to suppress ktlint rule must be fully qualified with the rule set id"),
+                ).isFormattedAs(formattedCode)
+        }
+
```
I am not in favor of testing all combinatoral combinations as that will really explode and will not add value.

paul-dingemans commented:
>

paul-dingemans commented:
> > @atulgpt I you also want to contribute by reviewing changes, you're welcome to review this PR. You can review it in detail or just briefly. Normally, I would just merge the change as there are almost no regular contributors on the project that would like to do this. I you don't want to review, just let me know.
> > REVIEW suggestion:
> > ```
> > * The functional change for is contained in [this commit](https://github.com/pinterest/ktlint/pull/2068/commits/428d52e637d931754b3756f8912f64cea884f8d5)
> > 
> > * Other commits are supportive to the change
> > ```
> 
> Hi, @paul-dingemans I have added a few more TC suggestions. Will take a look in the logic as well

Wow, I am really amazed at the good quality of the review comments. Tnx! I have left some review comments unanswered. I have to think about them.

paul-dingemans commented:
Can you please "resolve" the remarks on which I have responded and on which you agree? In that way, it is more clear for both of us what still needs to be fixed or discussed. Tnx. You're doing really great as reviewer!

atulgpt commented:
> Can you please "resolve" the remarks on which I have responded and on which you agree? In that way, it is more clear for both of us what still needs to be fixed or discussed. Tnx. You're doing really great as a reviewer!
<img width="893" alt="Screenshot 2023-06-08 at 1 38 49 AM" src="https://github.com/pinterest/ktlint/assets/11026203/b020a2a0-572b-474e-b436-249ddc73cd48">

Somehow it doesn't give me that option

paul-dingemans commented:
Hmm, I can not change that. If you leave a thumbs up or 'ok' message then I will resolve them.

paul-dingemans commented:
Hi @atulgpt ,
I hope you can find some time for review of logic.

atulgpt commented on ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/rules/KtlintSuppressionRule.kt:
```diff
@@ -0,0 +1,635 @@
+        when (ktlintDirectiveType) {
+            KTLINT_DISABLE -> {
+                if (node.elementType == EOL_COMMENT && node.prevLeaf().isWhiteSpaceWithNewline()) {
+                    removeDanglingEolCommentWithKtlintDisableDirective(autoCorrect, emit)
```
Somehow below TC is coming a dangling eol comment
```
val code =
    """
    class Foo() {
        val foo: Int
            // ktlint-disable standard:package-name
            get() = 1
    }
    """.trimIndent()
val formattedCode =
    """
    class Foo() {
        val foo1: Int
            @Suppress("ktlint:standard:package-name")
            get() = 1
    }
    """.trimIndent()
ktlintSuppressionRuleAssertThat(code)
    .isFormattedAs(formattedCode)
```

atulgpt commented:
Hi @paul-dingemans somehow the below TC is failing(it is adding multiple annotations)
```
val code =
    """
    class Foo() {
        val foo: Int
            get() = 1
            set(value: Int) { // ktlint-disable standard:foo
                field = value // ktlint-disable standard:foo
            }
    }
    """.trimIndent()
val formattedCode =
    """
    class Foo() {
        val foo: Int
            get() = 1
            @Suppress("ktlint:standard:foo")
            set(value: Int) {
                field = value
            }
    }
    """.trimIndent()
```

atulgpt commented:
> Hi @paul-dingemans somehow the below TC is failing(it is adding multiple annotations)
> 
> ```
> val code =
>     """
>     class Foo() {
>         val foo: Int
>             get() = 1
>             set(value: Int) { // ktlint-disable standard:foo
>                 field = value // ktlint-disable standard:foo
>             }
>     }
>     """.trimIndent()
> val formattedCode =
>     """
>     class Foo() {
>         val foo: Int
>             get() = 1
>             @Suppress("ktlint:standard:foo")
>             set(value: Int) {
>                 field = value
>             }
>     }
>     """.trimIndent()
> ```

```
class Foo() {
    var foo: Int = 0
        get() = 1
        @Suppress(names = ["abc"])
        set(value: Int) {
            field = value // ktlint-disable standard:foo
        }
}
``` 
this as well is failing

atulgpt commented:
> > Hi @paul-dingemans somehow the below TC is failing(it is adding multiple annotations)
> > ```
> > val code =
> >     """
> >     class Foo() {
> >         val foo: Int
> >             get() = 1
> >             set(value: Int) { // ktlint-disable standard:foo
> >                 field = value // ktlint-disable standard:foo
> >             }
> >     }
> >     """.trimIndent()
> > val formattedCode =
> >     """
> >     class Foo() {
> >         val foo: Int
> >             get() = 1
> >             @Suppress("ktlint:standard:foo")
> >             set(value: Int) {
> >                 field = value
> >             }
> >     }
> >     """.trimIndent()
> > ```
> 
> ```
> class Foo() {
>     var foo: Int = 0
>         get() = 1
>         @Suppress(names = ["abc"])
>         set(value: Int) {
>             field = value // ktlint-disable standard:foo
>         }
> }
> ```
> 
> this is also failing

I think currently we are also not handling type aliases like below
```
typealias A = Int // ktlint-disable standard:foo
```

atulgpt commented:
> > > Hi @paul-dingemans somehow the below TC is failing(it is adding multiple annotations)
> > > ```
> > > val code =
> > >     """
> > >     class Foo() {
> > >         val foo: Int
> > >             get() = 1
> > >             set(value: Int) { // ktlint-disable standard:foo
> > >                 field = value // ktlint-disable standard:foo
> > >             }
> > >     }
> > >     """.trimIndent()
> > > val formattedCode =
> > >     """
> > >     class Foo() {
> > >         val foo: Int
> > >             get() = 1
> > >             @Suppress("ktlint:standard:foo")
> > >             set(value: Int) {
> > >                 field = value
> > >             }
> > >     }
> > >     """.trimIndent()
> > > ```
> > 
> > 
> > ```
> > class Foo() {
> >     var foo: Int = 0
> >         get() = 1
> >         @Suppress(names = ["abc"])
> >         set(value: Int) {
> >             field = value // ktlint-disable standard:foo
> >         }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > this is also failing
> 
> I think currently we are also not handling type aliases like below
> 
> ```
> typealias A = Int // ktlint-disable standard:foo
> ```

And below code behaves strangely(maybe I am not setting the TC correctly)
```
val code =
    """
    private val a by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:foo
        val t = 0 // ktlint-disable standard:foo
        2
    }
    """.trimIndent()
val formattedCode =
    """
    private val a by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:foo
        val t = 0 // ktlint-disable standard:foo
        2
    }
    """.trimIndent()
```

above code doesn't show any violations but if I change the code like below
```
val code =
    """
    private val a by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:foo
        val t = 0 // ktlint-disable standard:foo
        2
    }
    fun t() { // ktlint-disable standard:foo

    }
    """.trimIndent()
```

it gets itself corrected as below code 
```
private val a by lazy(LazyThreadSafetyMode.PUBLICATION) @Suppress("ktlint:standard:foo")
  {
      @Suppress("ktlint:standard:foo")
      val t = 0
      2
  }
  @Suppress("ktlint:standard:foo")
  fun t() {
  
  }
```

paul-dingemans commented on documentation/snapshot/docs/faq.md:
```diff
@@ -43,21 +43,17 @@ See [adding a custom rule set](../api/custom-rule-set/) for more information.
+To disable a specific rule you'll need the fully qualified rule identifier. This identifier is displayed at the end of the lint error. In case your code was autocorrected, you need to revert the code and run the `lint` task instead of the `format` to find the rule identifier.
 
-An error can be suppressed using:
+As of Ktlint 0.50, an error can only be suppressed using @Suppress or @SuppressWarning annotations
```
I have found the same typo in other places as well. Fixed all of them.

paul-dingemans commented on documentation/snapshot/docs/rules/standard.md:
```diff
@@ -251,6 +251,54 @@ Indentation formatting - respects `.editorconfig` `indent_size` with no continua
 
+## Ktlint-suppression rule
+
+The `ktlint-disable` and `ktlint-enable` directives are no longer supported as of ktlint version `0.50.0`. This rule migrates the directives to Suppress or SuppressWarning annotations.
```
done

paul-dingemans commented on documentation/snapshot/docs/rules/standard.md:
```diff
@@ -251,6 +251,54 @@ Indentation formatting - respects `.editorconfig` `indent_size` with no continua
+Rule id: `ktlint-suppression` (`standard` rule set)
+
+!!! note
+    This rule can not be disabled in the `.editorconfig`.
```
The note states that this rule can *not* be disabled. So there is no property that can be used to disable it.

atulgpt replied:
Ohh I misread it 😅

paul-dingemans commented on ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/rulefilter/InternalRuleProvidersFilter.kt:
```diff
@@ -0,0 +1,38 @@
+    override fun filter(ruleProviders: Set<RuleProvider>): Set<RuleProvider> {
+        val internalRuleIds = internalRuleProviders.map { it.ruleId }
+        return ruleProviders
+            .mapNotNull {
+                if (it.ruleId in internalRuleIds) {
+                    LOGGER.error { "The provided rule with id '${it.ruleId}' is ignored in favour of Ktlint's rule with same id" }
+                    null
+                } else {
+                    it
+                }
+            }.plus(internalRuleProviders)
+            .toSet()
```
Done but it won't have any noticeable performance impact as this function is called once per file being scanned.

paul-dingemans commented on ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/rules/KtlintSuppressionRule.kt:
```diff
@@ -0,0 +1,635 @@
+            }
+            ?: false
+
+    private fun String.isKtlintSuppressionId() = removePrefix(DOUBLE_QUOTE).startsWith(KTLINT_SUPPRESSION_ID_PREFIX)
```
True but that is not very likely to be used in simple strings for the ids in `@Suppress`.

252K files found on github using `@Suppress("...")` => https://github.com/search?q=%40Suppress%28%22&type=code

39 occurrences found using `@Suppress(""` but interestingly enough this is always `@Suppress(""` and not `@Suppress("""...""")`.

atulgpt replied:
I think `@Suppress("")` mean something. There is one TC in detekt repo around `@Suppress("""...""")` at https://github.com/cortinico/detekt/blob/ebbb8402c856103dc7d95b192d9f03cb034acdd8/detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/StringShouldBeRawStringSpec.kt#L360 
strangely that is also written by me 🤣

Yes we can ignore it as no real usage is there

paul-dingemans commented on ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/rules/KtlintSuppressionRule.kt:
```diff
@@ -0,0 +1,635 @@
+    ) {
+        emit(offset, "Directive 'ktlint-disable' in EOL comment is ignored as it is not preceded by a code element", true)
+        if (autoCorrect) {
+            node
+                .prevLeaf()
+                .takeIf { it.isWhiteSpace() }
+                ?.remove()
```
Extracted to function `ASTNode.removePrecedingWhitespace()`

paul-dingemans commented:
> Hi @paul-dingemans somehow the below TC is failing(it is adding multiple annotations)
> 
> ```
> val code =
>     """
>     class Foo() {
>         val foo: Int
>             get() = 1
>             set(value: Int) { // ktlint-disable standard:foo
>                 field = value // ktlint-disable standard:foo
>             }
>     }
>     """.trimIndent()
> val formattedCode =
>     """
>     class Foo() {
>         val foo: Int
>             get() = 1
>             @Suppress("ktlint:standard:foo")
>             set(value: Int) {
>                 field = value
>             }
>     }
>     """.trimIndent()
> ```

Fixed

paul-dingemans commented:
> class Foo() {
>     var foo: Int = 0
>         get() = 1
>         @Suppress(names = ["abc"])
>         set(value: Int) {
>             field = value // ktlint-disable standard:foo
>         }
> }

I have simplified your example to:
```
            @Suppress(names = ["unused"])
            val foo = "foo" // ktlint-disable standard:foo
```

Fixed. When merging the annotations, I have chosen to drop the usage of the name argument for the sake of simplicity. So the result of above is:
```
            @Suppress("ktlint:standard:foo", "unused")
            val foo = "foo"
```

paul-dingemans commented:
> I think currently we are also not handling type aliases like below
> 
> ```
> typealias A = Int // ktlint-disable standard:foo
> ```

Type aliases are handled. No fix required.

paul-dingemans commented:
> And below code behaves strangely(maybe I am not setting the TC correctly)
> 
> ```
> val code =
>     """
>     private val a by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:foo
>         val t = 0 // ktlint-disable standard:foo
>         2
>     }
>     """.trimIndent()
> val formattedCode =
>     """
>     private val a by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:foo
>         val t = 0 // ktlint-disable standard:foo
>         2
>     }
>     """.trimIndent()
> ```
> 
> above code doesn't show any violations but if I change the code like below
> 
> ```
> val code =
>     """
>     private val a by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:foo
>         val t = 0 // ktlint-disable standard:foo
>         2
>     }
>     fun t() { // ktlint-disable standard:foo
> 
>     }
>     """.trimIndent()
> ```
> 
> it gets itself corrected as below code
> 
> ```
> private val a by lazy(LazyThreadSafetyMode.PUBLICATION) @Suppress("ktlint:standard:foo")
>   {
>       @Suppress("ktlint:standard:foo")
>       val t = 0
>       2
>   }
>   @Suppress("ktlint:standard:foo")
>   fun t() {
>   
>   }
> ```

Nice TC. I have use a simplified example to reproduce:
```
val foo by lazy(LazyThreadSafetyMode.PUBLICATION) { // ktlint-disable standard:indent
    // do something
}
```
which is now formatted as:
```
val foo by @Suppress("ktlint:standard:indent")
    lazy(LazyThreadSafetyMode.PUBLICATION) {
        // do something
    }
```

This example also learned that in case of nested expression, the annotation was not created at the outer most expression which resulted in code which could not be compiled. For example when using FUNCTION_LITERAL or LAMBDA_EXPRESSION:
```
val foo =
    setOf("a")
        .map {
            bar(it) // ktlint-disable standard:indent
        }

val foo = foo() {
    bar(it) // ktlint-disable standard:indent
}
```
This is also fixed.

atulgpt commented on ktlint-rule-engine/src/test/kotlin/com/pinterest/ktlint/rule/engine/internal/SuppressionLocatorBuilderTest.kt:
```diff
@@ -355,14 +355,15 @@ class SuppressionLocatorBuilderTest {
-                EMPTY_EDITOR_CONFIG_OVERRIDE
-                    .plus(
-                        STANDARD_NO_FOO_IDENTIFIER_RULE_ID.createRuleExecutionEditorConfigProperty() to RuleExecution.enabled,
+                ruleProviders =
```
@paul-dingemans under which rule this got formatted?

paul-dingemans replied:
I am not sure whether I understand.

But the IndentationRule would format:
```
                return mapOf(
                       1 to "   1 ms",
                      10 to "  10 ms",
                     999 to " 999 ms",
                    1000 to "   1 sec",
                )
```
to
```
                return mapOf(
                    1 to "   1 ms",
                    10 to "  10 ms",
                    999 to " 999 ms",
                    1000 to "   1 sec",
                )
```
if the rule would not be suppressed.

Does that answer the question?

atulgpt replied:
Actually, I am wondering why can't `ruleProviders = setOf(` be on the same line? Is this coz ktlint official recommends to always(in case of multiline chain) to newline after `=`
